neqo-http3/src/stream_type_reader.rs

```diff
@@ -82,3 +83,124 @@ impl RecvStream for NewStreamTypeReader {
+        decoder: NewStreamTypeReader,
+    }
+
+    impl Test {
+        fn new() -> Self {
+            let (mut conn_c, mut conn_s) = connect();
+            // create a stream
+            let stream_id = conn_s.stream_create(StreamType::UniDi).unwrap();
+            let out = conn_s.process(None, now());
+            mem::drop(conn_c.process(out.dgram(), now()));
+
+            Self {
+                conn_c,
+                conn_s,
+                stream_id,
+                decoder: NewStreamTypeReader::new(stream_id),
+            }
+        }
+
+        fn decode(&mut self, stream_type: u64, outcome: ReceiveOutput, done: bool) {
+            let mut enc = Encoder::default();
+            enc.encode_varint(stream_type);
+            let len = enc.len() - 1;
+            for i in 0..len {
+                self.conn_s
+                    .stream_send(self.stream_id, &enc[i..i + 1])
+                    .unwrap();
+                let out = self.conn_s.process(None, now());
+                mem::drop(self.conn_c.process(out.dgram(), now()));
+                assert_eq!(
+                    self.decoder.receive(&mut self.conn_c).unwrap(),
+                    ReceiveOutput::NoOutput
+                );
+                assert_eq!(self.decoder.done(), false);
+            }
+            self.conn_s
+                .stream_send(self.stream_id, &enc[enc.len() - 1..])
+                .unwrap();
+            let out = self.conn_s.process(None, now());
+            mem::drop(self.conn_c.process(out.dgram(), now()));
+            assert_eq!(self.decoder.receive(&mut self.conn_c).unwrap(), outcome);
+            assert_eq!(self.decoder.done(), done);
+        }
+    }
+
+    #[test]
+    fn decode_streams() {
+        let mut t = Test::new();
+        t.decode(
+            QPACK_UNI_STREAM_TYPE_DECODER,
+            ReceiveOutput::NewStream(Http3StreamType::Encoder),
+            true,
+        );
+
+        let mut t = Test::new();
+        t.decode(
+            QPACK_UNI_STREAM_TYPE_ENCODER,
+            ReceiveOutput::NewStream(Http3StreamType::Decoder),
+            true,
+        );
+
+        let mut t = Test::new();
+        t.decode(
+            HTTP3_UNI_STREAM_TYPE_CONTROL,
+            ReceiveOutput::NewStream(Http3StreamType::Control),
+            true,
+        );
+
+        let mut t = Test::new();
+        t.decode(
+            HTTP3_UNI_STREAM_TYPE_PUSH,
+            ReceiveOutput::NewStream(Http3StreamType::Push),
+            true,
+        );
+
+        let mut t = Test::new();
+        t.decode(
+            0x3fff_ffff_ffff_ffff,
+            ReceiveOutput::NewStream(Http3StreamType::Unknown),
+            true,
+        );
+    }
+
+    #[test]
+    fn done_decoding() {
+        let mut t = Test::new();
+        t.decode(
+            0x3fff,
+            ReceiveOutput::NewStream(Http3StreamType::Unknown),
+            true,
+        );
+        t.decode(HTTP3_UNI_STREAM_TYPE_PUSH, ReceiveOutput::NoOutput, true);
```

Comment:
I think that you need to include the truncated identifier option also.  That means literally encoding 0xff and nothing more.  Maybe the way to do that is to pass a pre-encoded byte array for the identifier.